src/torch_models.py

```diff
@@ -295,305 +502,102 @@ def predict_sample(self, x: Array, rng: np.random.Generator) -> Array:
-        # Train
-        logging.info(f"Training {self.__class__.__name__} on {X.shape[0]} "
-                     "datapoints")
-        self.train()  # switch to train mode
-        itr = 0
-        best_loss = float("inf")
-        best_itr = 0
-        n_iter_no_change = CFG.mlp_classifier_n_iter_no_change
-        model_name = tempfile.NamedTemporaryFile(delete=False).name
-        loss_fn = nn.BCELoss()
-        optimizer = optim.Adam(self.parameters(), lr=CFG.learning_rate)
-        while True:
-            yhat = self(X)
-            loss = loss_fn(yhat, y)
-            if loss.item() < best_loss:
-                best_loss = loss.item()
-                best_itr = itr
-                # Save this best model
-                torch.save(self.state_dict(), model_name)
-            if itr % 1000 == 0:
-                logging.info(f"Loss: {loss:.5f}, iter: {itr}/{self._max_itr}")
-            optimizer.zero_grad()
-            loss.backward()
-            optimizer.step()
-            if itr == self._max_itr:
-                break
-            if itr - best_itr > n_iter_no_change:
-                logging.info(f"Loss did not improve after {n_iter_no_change} "
-                             f"itrs, terminating at itr {itr}.")
-                break
-            itr += 1
-        # Load best model
-        self.load_state_dict(torch.load(model_name))  # type: ignore
-        os.remove(model_name)
-        self.eval()  # switch to eval mode
-        yhat = self(X)
-        loss = loss_fn(yhat, y)
-        logging.info(f"Loaded best model with loss: {loss:.5f}")
-
-
-class MLPClassifierEnsemble(Classifier):
+class MLPClassifierEnsemble(BinaryClassifier):
     """MLPClassifierEnsemble definition."""
 
-    def __init__(self, in_size: int, max_itr: int, n: int) -> None:
+    def __init__(self, seed: int, balance_data: bool, max_train_iters: int,
+                 learning_rate: float, n_iter_no_change: int,
+                 hid_sizes: List[int], ensemble_size: int) -> None:
+        super().__init__(seed, balance_data)
         self._members = [
-            MLPClassifier(in_size, max_itr, CFG.seed + i) for i in range(n)
+            MLPClassifier(seed + i, balance_data, max_train_iters,
+                          learning_rate, n_iter_no_change, hid_sizes)
+            for i in range(ensemble_size)
         ]
 
     def fit(self, X: Array, y: Array) -> None:
+        # Each member maintains its own normalizers.
         for i, member in enumerate(self._members):
             logging.info(f"Fitting member {i} of ensemble...")
             member.fit(X, y)
 
     def classify(self, x: Array) -> bool:
+        # Each member maintains its own normalizers.
         avg = np.mean(self.predict_member_probas(x))
         classification = avg > 0.5
         assert classification in [True, False]
         return classification
 
+    def _fit(self, X: Array, y: Array) -> None:
+        raise NotImplementedError("Not used.")
```

Comment:
maybe change this to `raise Exception` since it never makes sense to implement it?